torchtitan/optimizer.py

```diff
@@ -79,27 +118,50 @@ def linear_warmup_linear_decay(
     return curr_adjustment
 
 
+class SchedulersContainer:
+    """Util for calling step on multiple learning rate schedulers needed for virtual pipeline stages"""
+
+    def __init__(self, optimizers, lr_lambda):
+        self.schedulers = []
+        for optimizer in optimizers:
+            self.schedulers.append(LambdaLR(optimizer, lr_lambda=lr_lambda))
+
+    def step(self):
+        for schedulers in self.schedulers:
+            schedulers.step()
+
+
+class SchedulersInBackwardContainer(SchedulersContainer):
+    """Util for calling step on multiple learning rate schedulers when optimizers are in backward"""
+
+    def __init__(self, optimizers, lr_lambda):
+        # all the schedulers for each optimizer group are the same, here we only store the first one
+        # to self.schedulers follow the same structure as SchedulersContainer, but store all of them
+        # to self.all_schedulers for container.step() to call
+        self.schedulers = []
+        self.all_schedulers = []
+        for optim_group in optimizers:
+            if len(optim_group) > 0:
+                self.schedulers.append(LambdaLR(optim_group[0], lr_lambda=lr_lambda))
```

Comment:
this doesn't sound right -- It seems what you store in the `state_dict` is not the one used in training. E.g. after several steps, the scheduler in training would decrease the learning rate, but the one you created here would still have original LR.

I still think we should address this by unifying "container" and "wrapper". This way the `state_dict` of `SchedulersInBackwardContainer` can just save the `self.all_schedulers[0][0]`.

I'm OK with having a separate PR for it, but here you'll need to change `self.schedulers` to use the objects in `self.all_schedulers` rather than creating its own.

Reply:
Yeah, the logic here is not right, temporally using `self.all_schedulers[0][0]` like expression under isinstance(lr_schedulers, SchedulersInBackwardContainer).
Would have a separate PR to restructure optimzierwrapper and schedulerwrapper.